onnxruntime/python/tools/transformers/convert_generation.py

```diff
@@ -922,13 +965,29 @@ def convert_generation_model(args: argparse.Namespace, generation_type: Generati
         logits_matmul_weight_padded = pad_weights_of_logits_matmul(args.decoder_onnx, args.use_external_data_format)
         if not logits_matmul_weight_padded:
             logger.warning(
-                "Tried and failed to pad logits MatMul weights. " "Performance may be sub-optimal for this MatMul"
+                "Tried and failed to pad logits MatMul weights. Performance may be sub-optimal for this MatMul"
+            )
+
+    # We only want to project the encoding of the last token in the batched sequence
+    # onto the vocabulary space. We hence replace the MatMul operator with a customized
+    # operator that does just that.
+    # NOTE: We currently only support this for fp16/fp32 GPT2 BeamSearch.
+    # This can be expanded to other models/decoding strategies/data types later
+    last_token_logits_matmul_enabled = False
+    if args.enable_last_token_logits_matmul and is_gpt2 and generation_type == GenerationType.BEAMSEARCH:
+        logger.info(
+            f"Replace logits MatMul with LastTokenMatMul on {args.decoder_onnx}. " "The file will be overwritten."
+        )
+        last_token_logits_matmul_enabled = replace_logits_matmul(args.decoder_onnx, args.use_external_data_format)
```

Comment:
For GPT-2, the first decoding is done in decoder_onnx. For encoder-decoder model like T5, first decoding is done in encoder_decoder_init subgraph.

Reply:
For now, this method is only called when the model is GPT2 (and so was the previous optimization wrt to vocab padding). Would you like to expand this for T5 or prefer that we do it in a separate change later ?